Initial modules documentation (#1327)
hhvm/user-documentation: guides/hack/56-modules/06-type-aliases.md

Type aliases have special rules when it comes to internal types depending on which kind of type alias is used. 

## Internal type aliases

> An opaque or normal type alias can be marked internal. Just as with classes, this means that the type can only be referenced from within the module.

Reviewer: I'd use concrete syntax rather than saying 'opaque or normal', as you're referring to a single kind of type type alias.

```suggestion
A type alias declared with `type` can be marked internal. Just as with classes, this means that the type can only be referenced from within the module.
```

```hack
newtype FooOpaque = FooInternal; // ok 
newtype FooErr as FooInternal = FooInternal; // error, FooInternal is an internal type, cannot be used as constraint
```

## Public transparent type aliases
Transparent type aliases leak their implementation, so they cannot be implemented by internal types. 
```
type FooOpaque = FooInternal; // error, cannot use internal type FooInternal in transparent type FooOpaque.
```

## Module-level type aliases
We also introduce a new type alias known as a **module-level** type alias. You can create one using the syntax `module newtype TFoo as ... = ...`.
A module-level type alias is opaque outside of the module it's defined in, and transparent inside (rather than just being opaque outside of the file it's defined in). Since they hide their implementations from outside the module, you can use internal types from within a module to implement them. They still cannot be constrained by internal types. 

```hack
module newtype FooModule = FooInternal; // ok 
module newtype FooModuleErr as FooInternal = FooInternal; // error, FooInternal is an internal type, cannot be used as a constraint 
```

> Since the entire point of module newtypes is to create an interface surrounding a module boundary, you cannot mark module newtypes themselves internal. 

Reviewer: ```suggestion
Since the purpose of module newtypes is to create an interface surrounding a module boundary, you cannot mark module newtypes themselves internal. 
```

'The entire point' implies the user is being a little dumb to me.
